File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/items/media/image/specs/GivenAServiceFile/WhenGettingTheImageBytes.java
package com.lasthopesoftware.bluewater.client.browsing.items.media.image.specs.GivenAServiceFile;

import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.browsing.items.media.image.RemoteImageAccess;
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId;
import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider;
import com.lasthopesoftware.bluewater.client.connection.specs.FakeConnectionProvider;
import com.lasthopesoftware.bluewater.client.connection.specs.FakeConnectionResponseTuple;
import com.lasthopesoftware.bluewater.client.connection.specs.FakeLibraryConnectionProvider;
import com.lasthopesoftware.bluewater.shared.promises.extensions.specs.FuturePromise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.HashMap;
import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.Assertions.assertThat;

public class WhenGettingTheImageBytes {

	private static byte[] imageBytes;

	@BeforeClass
	public static void before() throws ExecutionException, InterruptedException {
		final FakeConnectionProvider fakeConnectionProvider = new FakeConnectionProvider();
		fakeConnectionProvider.mapResponse(
			p -> new FakeConnectionResponseTuple(200, new byte[] { 39, 127, 8 }),
			"File/GetImage", "File=31", "Type=Full", "Pad=1", "Format=jpg", "FillTransparency=ffffff");

		final RemoteImageAccess memoryCachedImageAccess = new RemoteImageAccess(
			new FakeLibraryConnectionProvider(
				new HashMap<LibraryId, IConnectionProvider>() {
					{
						put(new LibraryId(21), fakeConnectionProvider);
					}
				}
			)
		);

		imageBytes = new FuturePromise<>(memoryCachedImageAccess.promiseImageBytes(
			new LibraryId(21), new ServiceFile(31))).get();
	}

	@Test
	public void thenTheBytesAreCorrect() {
		assertThat(imageBytes).isEqualTo(new byte[] { 39, 127, 8 });
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/browsing/items/media/image/GivenAServiceFile/WhenGettingTheImageBytes.kt
package com.lasthopesoftware.bluewater.client.browsing.items.media.image.GivenAServiceFile

import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.items.media.image.RemoteImageAccess
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.FakeConnectionProvider
import com.lasthopesoftware.bluewater.client.connection.FakeConnectionResponseTuple
import com.lasthopesoftware.bluewater.client.connection.FakeLibraryConnectionProvider
import com.lasthopesoftware.bluewater.shared.promises.extensions.toFuture
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class WhenGettingTheImageBytes {
	companion object {
		private val imageBytes by lazy {
			val fakeConnectionProvider = FakeConnectionProvider()
			fakeConnectionProvider.mapResponse(
				{ FakeConnectionResponseTuple(200, byteArrayOf(39, 127, 8)) },
				"File/GetImage",
				"File=31",
				"Type=Full",
				"Pad=1",
				"Format=jpg",
				"FillTransparency=ffffff"
			)

			val memoryCachedImageAccess = RemoteImageAccess(
				FakeLibraryConnectionProvider(mapOf(Pair(LibraryId(21), fakeConnectionProvider))))

			memoryCachedImageAccess.promiseImageBytes(LibraryId(21), ServiceFile(31)).toFuture().get()
		}
	}

	@Test
	fun thenTheBytesAreCorrect() {
		assertThat(imageBytes).isEqualTo(byteArrayOf(39, 127, 8))
	}
}
